modules/edu.gemini.seqexec.engine/src/main/scala/edu/gemini/seqexec/engine/Engine.scala

```diff
@@ -3,36 +3,374 @@
+      case GetSeqState(id, f)         => getSeqState(id, f)
+      case ActionStop(id, f)          => Logger.debug("Engine: Action stop requested") *> actionStop(id, f)
+      case ActionResume(id, i, cont)  => Logger.debug("Engine: Action resume requested") *> actionResume(id, i, cont)
+      case LogDebug(msg)              => Logger.debug(msg)
+      case LogInfo(msg)               => Logger.info(msg)
+      case LogWarning(msg)            => Logger.warning(msg)
+      case LogError(msg)              => Logger.error(msg)
+    }
+
+    def handleSystemEvent(se: SystemEvent): HandleP[Unit, D] = se match {
+      case Completed(id, i, r)     => Logger.debug("Engine: Action completed") *> complete(id, i, r)
+      case PartialResult(id, i, r) => Logger.debug("Engine: Partial result") *> partialResult(id, i, r)
+      case Paused(id, i, r)        => Logger.debug("Engine: Action paused") *> actionPause(id, i, r)
+      case Failed(id, i, e)        => logError(e) *> fail(id)(i, e)
+      case Busy(id)                => Logger.warning(s"Cannot run sequence $id because required systems are in use.")
+      case BreakpointReached(_)    => Logger.debug("Engine: Breakpoint reached")
+      case Executed(id)            => Logger.debug("Engine: Execution completed") *> next(id)
+      case Executing(id)           => Logger.debug("Engine: Executing") *> execute(id)
+      case Finished(id)            => Logger.debug("Engine: Finished") *> switch(id)(SequenceState.Completed)
+      case Null                    => unit
+    }
+
+    (ev match {
+      case EventUser(ue)   => handleUserEvent(ue)
+      case EventSystem(se) => handleSystemEvent(se)
+    }) *> get
+  }
+
+  // Kudos to @tpolecat
+  /** Traverse a process with a stateful computation. */
+  private def mapEvalState[A, S, B](
+                                     fs: Process[Task, A], s0: S, f: A => StateT[Task, S, (B, Option[Process[Task, A]])]
+                                   ): Process[Task, B] = {
+    def go(fi: Process[Task, A], si: S): Process[Task, B] = {
+      Process.eval(fi.unconsOption).flatMap {
+        case None => Process.halt
+        case Some((h, t)) => Process.eval(f(h).run(si)).flatMap {
+          case (s, (b, p)) => Process.emit(b) ++ go(p.map(_ merge t).getOrElse(t), s)
+        }
+      }
+    }
+
+    go(fs, s0)
+  }
+
+  private def runE(ev: Event[D]): HandleP[(Event[D], Engine.State[D]), D] =
+    run(ev).map((ev, _))
+
+  def process(input: Process[Task, Event[D]])(qs: Engine.State[D]): Process[Task, (Event[D], Engine.State[D])] =
+    mapEvalState[Event[D], Engine.State[D], (Event[D], Engine.State[D])](input, qs, (e: Event[D]) => runE(e).run)
+
+  // Functions for type bureaucracy
+
+  private def pure[A](a: A): HandleP[A, D] = Applicative[({type L[T] = HandleP[T, D]})#L].pure(a)
+
+  private val unit: HandleP[Unit, D] = pure(())
+
+  private val get: HandleP[Engine.State[D], D] =
+    MonadState[({type L[T] = Handle[T, D]})#L, Engine.State[D]].get.toHandleP
```

Comment:
You may not need `MonadState` here. It should be possible to do the `get`, `modify` calls directly on `StateT`